sqlx-core/src/postgres/types/decimal.rs

```diff
@@ -0,0 +1,443 @@
+    fn type_info() -> PgTypeInfo {
+        <[Decimal] as Type<Postgres>>::type_info()
+    }
+}
+
+impl TryFrom<PgNumeric> for Decimal {
+    type Error = BoxDynError;
+
+    fn try_from(numeric: PgNumeric) -> Result<Self, BoxDynError> {
+        let (digits, sign, weight) = match numeric {
+            PgNumeric::Number {
+                digits,
+                sign,
+                weight,
+                ..
+            } => (digits, sign, weight),
+
+            PgNumeric::NotANumber => {
+                return Err("Decimal does not support NaN values".into());
+            }
+        };
+
+        if digits.is_empty() {
+            // Postgres returns an empty digit array for 0 but BigInt expects at least one zero
+            return Ok(0u64.into());
+        }
+
+        let sign = match sign {
+            PgNumericSign::Positive => Sign::Plus,
+            PgNumericSign::Negative => Sign::Minus,
+        };
+
+        // weight is 0 if the decimal point falls after the first base-10000 digit
+        let scale = (digits.len() as i64 - weight as i64 - 1) * 4;
+
+        // no optimized algorithm for base-10 so use base-100 for faster processing
+        let mut cents = Vec::with_capacity(digits.len() * 2);
+        for digit in &digits {
+            cents.push((digit / 100) as u8);
+            cents.push((digit % 100) as u8);
+        }
+
+        let bigint = BigInt::from_radix_be(sign, &cents, 100)
+            .ok_or("PgNumeric contained an out-of-range digit")?;
+
+        match bigint.to_i128() {
+            Some(num) => Ok(Decimal::from_i128_with_scale(num, scale as u32)),
+            None => Err("Decimal's integer part out of range.".into()),
+        }
+    }
+}
+
+impl TryFrom<&'_ Decimal> for PgNumeric {
+    type Error = BoxDynError;
+
+    fn try_from(decimal: &Decimal) -> Result<Self, BoxDynError> {
+        if decimal.is_zero() {
+            return Ok(PgNumeric::Number {
+                sign: PgNumericSign::Positive,
+                scale: 0,
+                weight: 0,
+                digits: vec![],
+            });
+        }
+
+        let scale = decimal.scale() as u16;
+        let groups_diff = scale & 0x3;
+
+        // A serialized version of the decimal number. The resulting byte array
+        // will have the following representation:
+        //
+        // Bytes 1-4: flags
+        // Bytes 5-8: lo portion of m
+        // Bytes 9-12: mid portion of m
+        // Bytes 13-16: high portion of m
+        let s = decimal.serialize();
+
+        let mut mantissa = [
+            u32::from_le_bytes(s[4..8].try_into().unwrap()),
+            u32::from_le_bytes(s[8..12].try_into().unwrap()),
+            u32::from_le_bytes(s[12..16].try_into().unwrap()),
+            0u32,
+        ];
+
+        if groups_diff > 0 {
+            let remainder: u16 = 4 - groups_diff;
+            let power = 10u32.pow(remainder as u32);
+            mul_by_u32(&mut mantissa, power);
+        }
```

Comment:
I just realized that implementing this manually is pointless; all you need to do is wrap `scale` to the next multiple of 4 and use `Decimal::rescale()` by doing the following _before_ `decimal.serialize()`:

```rust
let scale = decimal.scale();
// if `scale` is not a multiple of 4, make it the *next* multiple of 4
// by adding 3, if `scale` is not a multiple of 4 then we're rounding it up to *at least* the next multiple of 4
// then by masking off the lower 2 bits we're effectively subtracting off however much extra over the multiple of 4
// if `scale` is a multiple of 4 then this is a no-op
let new_scale = scale + 3 & !3;

let decimal = decimal.rescale(new_scale);
```

Reply:
`rescale()` does have the caveat that it will _round_ if the scale is beyond the maximum precision (if all 96 bits are being used, there's no space to multiply by 10).

The extracted implementation doesn't appear to have the same issue because it adds the extra u32.